subsys/power/policy/policy_residency.c

```diff
@@ -72,9 +72,8 @@ int sys_pm_policy_next_state(s32_t ticks, enum power_states *pm_state)
 		return SYS_PM_NOT_HANDLED;
 	}
 
-	for (i = 0; i < (ARRAY_SIZE(pm_policy) - 1); i++) {
-		if ((ticks >= pm_policy[i].min_residency) &&
-			       (ticks < pm_policy[i + 1].min_residency)) {
+	for (i = ARRAY_SIZE(pm_policy) - 1; i >= 0; i--) {
+		if (ticks >= pm_policy[i].min_residency) {
```

Comment:
If ticks does not match any check look will end with `i` equal to -1 causing an overflow on the `pm_policy`.

In the old code cases when ticks do not match any region will end up with last policy value (this I expect is also incorrect).

Reply:
The if above loop handles that case.

Reply:
But for `ticks == K_FOREVER` the loop can end up with `i` equal to `-1`. The previous implementation would set the last policy for `K_FOREVER` ticks.

Reply:
Fixed. Thank you @anangl.